### DISCUSSION TITLE
Chained Control Flow Statements: break [, break]... [,continue]

### DISCUSSION THREAD

**Discussion by Caracrist** [dotnet/csharplang#8434] [Language Ideas] [answered]
The concept is to allow multiple `break` statements and an optional `continue` operation within a single line, as shown below:

```
while (cond1) // returns here from: break, continue
{
    while (cond2)
    {
        if (cond3)
        {
            break, break; // break from the inner and outer loop
        }
        else if (cond4)
        {
            break, continue; // break from the inner loop and continue in the outer loop
        }
    }
    // this code is reached only if !cond2
}
// code reached from "break, break" or if !cond1
```
```
while (cond1) // returns here from: break, continue
{
    switch (item)
    {
        case value1:
            break, break; // break from the switch and break outer loop
        case value2:
            break, continue; // break from the switch and continue in the outer loop
        default:
            break;
    }
    // this code is reached only if default
}
// code reached from "break, break" or if !cond1
```

The `break` statements can exit loops or `switch` case blocks. The number of `break` statements is limited by the number of enclosing scopes, while the `continue` statement can only appear once and must be the last in the sequence.

The formal definition is (see subj):

```
break [, break]... [, continue]
```

This means that within a block of code, multiple `break` statements and an optional `continue` can be used at the end. None are mandatory, ensuring full backward compatibility.

Typically, alternatives to this approach involve managing multiple flags, akin to a state machine spread across local variables, combined with many `if` statements to handle conditions in the right place. This can complicate both the code and its readability.

Additionally, since each statement (`break`, `continue`) is a distinct word, it can be highlighted in the IDE, showing its relation to the corresponding block (`for`, `foreach`, `while`, `switch`, `do`). During debugging, these statements can be executed one by one, similar to how execution is handled inside `for (statements)`, with each part of the line being highlighted independently.

**Comment by zhenkas** [comment]
Elegant! I want it in c++ too!

**Reply by emamontov** [reply]
This kind of feature that everybody need, but nobody knows that. 
+Upvoted

**Comment by HaloFour** [comment]
See:

#6634

**Reply by Caracrist** [reply]
Seen it, thank you.

All these alternatives are messy, and just replace flags management with labels and loop names management. Also none of these support the continue option and debugging each break one by one. For instance if there is a using statement in the inner loop, stopping on the second break/continue would allow you to see the state after dispose and before the next break/continue. It is also very clear solution with no names, flags and labels introduction into the code.

In fact, it is the lowest complexity solution possible.

Yet thank you @HaloFour for the link and the invitation to address it.

**Reply by timcassell** [reply]
I do prefer labeled loops (that would also work with continue).

**Reply by emamontov** [reply]
There is a discussion while here is the proposal in completed form.

**Reply by HaloFour** [reply]
I personally prefer the labels, it's very clear exactly where the flow will go.  `break, break;` is just as brittle as `break 2;` when it comes to potential refactoring.

Either way, the language team hasn't expressed much interest here as `goto` with a label is sufficient.

**Reply by VaronisContributor** [reply]
`goto` with a label is anti-pattern ...

**Reply by koszeggy** [reply]
> `goto` with a label is anti-pattern ...

...which is a big misunderstanding. `goto` can be dangerous in C/C++ but in C# it is much more restricted (e.g. you cannot just jump out of a `finally` block, or to code points where variables would not be initialized, not mentioning jumping into another method, or doing other crazy things that could even corrupt the stack).

The runtime itself also [uses](https://github.com/dotnet/runtime/blob/550d381a22796f4d3be3a3cd7ad5417db2cbe80f/src/libraries/System.Private.CoreLib/src/System/Marvin.cs#L36) [a](https://github.com/dotnet/runtime/blob/550d381a22796f4d3be3a3cd7ad5417db2cbe80f/src/libraries/System.Private.CoreLib/src/System/MemoryExtensions.Trim.cs#L737) [lot](https://github.com/dotnet/runtime/blob/550d381a22796f4d3be3a3cd7ad5417db2cbe80f/src/libraries/System.Private.CoreLib/src/System/String.Comparison.cs#L137) [of](https://github.com/dotnet/runtime/blob/550d381a22796f4d3be3a3cd7ad5417db2cbe80f/src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/TraceLogging/ConcurrentSet.cs#L111) [`goto`](https://github.com/dotnet/runtime/blob/550d381a22796f4d3be3a3cd7ad5417db2cbe80f/src/libraries/System.Private.CoreLib/src/System/SpanHelpers.cs#L50), especially on performance critical paths.

When that's the best tool, just use it. Don't afraid of others who have gotophobia.

**Comment by CyrusNajmabadi** [comment, marked as answer]
Proposal started here: https://github.com/dotnet/csharplang/blob/main/proposals/labeled-break-continue.md
Championed Issue here: https://github.com/dotnet/csharplang/issues/9875
Discussion here: https://github.com/dotnet/csharplang/discussions/9876

Closing this out to move discussion there.
